// turborepo-remote-cache@2.2.8/dist/plugins/remote-cache/storage/google-cloud-storage.js
import GCS from '@google-cloud/storage';
function createStorage({ clientEmail, privateKey, projectId }) {
    if (!clientEmail || !privateKey || !projectId) {
        return new GCS.Storage();
    }
    else {
        return new GCS.Storage({
            projectId,
            credentials: {
                client_email: clientEmail,
                private_key: privateKey,
            },
        });
    }
}
export function createGoogleCloudStorage({ bucket, clientEmail, privateKey, projectId, }) {
    const storage = createStorage({
        clientEmail,
        privateKey,
        projectId,
    });
    const turboBucket = storage.bucket(bucket);
    return {
        exists: (artifactPath, cb) => {
            turboBucket.file(artifactPath).exists(cb);
        },
        createReadStream(artifactPath) {
            return turboBucket
                .file(artifactPath)
                .createReadStream();
        },
        createWriteStream(artifactPath) {
            return turboBucket.file(artifactPath).createWriteStream();
        },
    };
}
