// turborepo-remote-cache@2.2.8/dist/plugins/remote-cache/storage/index.js
import { join } from 'path';
import { pipeline as pipelineCallback } from 'stream';
import { promisify } from 'util';
import { STORAGE_PROVIDERS } from '../../../env.js';
import { createAzureBlobStorage, } from './azure-blob-storage.js';
import { createGoogleCloudStorage, } from './google-cloud-storage.js';
import { createLocal } from './local.js';
import { createS3 } from './s3.js';
const pipeline = promisify(pipelineCallback);
const TURBO_CACHE_FOLDER_NAME = 'turborepocache';
const TURBO_CACHE_USE_TMP_FOLDER = true;
function createStorageLocation(provider, providerOptions) {
    const { path = TURBO_CACHE_FOLDER_NAME, useTmp = TURBO_CACHE_USE_TMP_FOLDER, } = providerOptions;
    switch (provider) {
        case STORAGE_PROVIDERS.LOCAL: {
            return createLocal({ path, useTmp });
        }
        case STORAGE_PROVIDERS.S3:
        case STORAGE_PROVIDERS.s3: {
            const { accessKey, secretKey, region, endpoint } = providerOptions;
            return createS3({ accessKey, secretKey, bucket: path, region, endpoint });
        }
        case STORAGE_PROVIDERS.MINIO: {
            const { accessKey, secretKey, region, endpoint } = providerOptions;
            return createS3({
                accessKey,
                secretKey,
                bucket: path,
                region,
                endpoint,
                s3OptionsPassthrough: {
                    s3ForcePathStyle: true,
                },
            });
        }
        case STORAGE_PROVIDERS.GOOGLE_CLOUD_STORAGE: {
            const { clientEmail, privateKey, projectId } = providerOptions;
            return createGoogleCloudStorage({
                bucket: path,
                clientEmail,
                privateKey,
                projectId,
            });
        }
        case STORAGE_PROVIDERS.AZURE_BLOB_STORAGE: {
            const { connectionString } = providerOptions;
            return createAzureBlobStorage({ containerName: path, connectionString });
        }
        default:
            throw new Error(`Unsupported storage provider '${provider}'. Please select one of the following: ${Object.values(STORAGE_PROVIDERS).join(', ')}!`);
    }
}
export function createLocation(provider, providerOptions) {
    const location = createStorageLocation(provider, providerOptions);
    async function getCachedArtifact(artifactId, team) {
        return new Promise((resolve, reject) => {
            const artifactPath = join(team, artifactId);
            location.exists(artifactPath, (err, exists) => {
                if (err) {
                    return reject(err);
                }
                if (!exists) {
                    return reject(new Error(`Artifact ${artifactPath} doesn't exist.`));
                }
                resolve(location.createReadStream(artifactPath));
            });
        });
    }
    async function existsCachedArtifact(artifactId, team) {
        return new Promise((resolve, reject) => {
            const artifactPath = join(team, artifactId);
            location.exists(artifactPath, (err, exists) => {
                if (err) {
                    return reject(err);
                }
                if (!exists) {
                    return reject(new Error(`Artifact ${artifactPath} doesn't exist.`));
                }
                resolve();
            });
        });
    }
    async function createCachedArtifact(artifactId, team, artifact) {
        return pipeline(artifact, location.createWriteStream(join(team, artifactId)));
    }
    return {
        getCachedArtifact,
        createCachedArtifact,
        existsCachedArtifact,
    };
}
